v1/dictionary/handlers.js

```diff
@@ -0,0 +1,133 @@
+'use strict';
+
+var fluid = require('infusion');
+var AdaptiveContentServices = fluid.registerNamespace('AdaptiveContentServices');
+var wd = require('word-definition');
+
+require('kettle');
+
+
+/* General function to get definition 
+ * from all the different third-party dictionary services being used
+ * Simply gives the complete response from the third party service 
+ * Does not manipulate it in any way
+ * The function returns a promise whose result contains the response
+ * from the third-party service
+ * Service codes -
+ *  Wiktionary -> 'wiki'
+ */
+function definition(service, lang, word) {
+
+  //Wiktionary Service
+  if(service == 'wiki') {
+    var promise = fluid.promise();
+    wd.getDef(word, lang, null, function(data) {
+      promise.resolve(data);
+    });
+    return promise;
+  }
+}
+
+/* General function to catch the errors
+ * From the third-party dictionary services
+ * Service codes -
+ *  Wiktionary -> 'wiki'
+ */
+function checkDictionaryErr(service, serviceResponse) {
+  //Wiktionary Service
+  if(service == 'wiki') {
+
+    //Check if there is an error
+    if(serviceResponse.err) {
+
+      //Word not found
+      if(serviceResponse.err == 'not found') {
+        return {
+          statusCode: 404,
+          errorMessage: 'Dictionary Service: Word not found (Wiktionary)'
+        }
+      }
+
+      //Language unsupported by the third-party service
+      else if(serviceResponse.err == 'unsupported language') {
+        return {
+          statusCode: 400,
+          errorMessage: 'Dictionary Service: Unsupported Language (Wiktionary)'
+        }
+      }
+
+      //Default return object when error hasn't been handled yet
+      else return {
+        statusCode: 500,
+        errorMessage: 'Dictionary Service: The error hasn\'t been handled yet (Wiktionary)'
+      };
+    }
+
+    //Return false if no error found
+    else return;
+  }
+}
+
+//Main request handler - Gives only the definition of the word
+fluid.defaults('AdaptiveContentServices.Dictionary.serverConfig.mainDictionaryHandler', {
```

Comment:
As I mentioned when we talked last Thursday, we should refactor this implementation to take advantage of specific features in Kettle / Infusion that support separating public interfaces from their implementations in configuration. This implementation is on the right track for supporting this by passing a "service" argument in its functions, but there's some specific things we can leverage that will let us build the services more flexibly, and eventually let us deploy customized sets of services fairly easily.

A basic form of this in a Kettle handler could look something like the following:

```
// First, we define an "abstract" grade that other
// grades we create can use; we define the public
// API here
fluid.defaults("example.handler.dictionaryDefinition", {
  gradeNames: ["kettle.request.http"],
  // This is a function in the core API that we
  // use to indicate a grade is abstract and
  // is intended to provide common material
  // (public API, etc) for other grades
  invokers: {
      handleRequest: "fluid.notImplemented"
  }
});

// Handler that uses wiktionary
fluid.defaults("example.handler.dictionaryDefinition.wiktionary", {
  gradeNames: ["example.handler.dictionaryDefinition"],
  invokers: {
      handleRequest: "example.handler.dictionaryDefinition.wiktionary.getDefinition"
  }
});

example.handler.dictionaryDefinition.wiktionary.getDefinition = function (request) {
    // Specific function implementation for Wiktionary here
};

// Handler that uses Oxford
fluid.defaults("example.handler.dictionaryDefinition.oxford", {
  gradeNames: ["example.handler.dictionaryDefinition"],
  invokers: {
      handleRequest: "example.handler.dictionaryDefinition.oxford.getDefinition"
  }
});

example.handler.dictionaryDefinition.oxford.getDefinition = function (request) {
    // Specific function implementation for Oxford here
};
```
The advantage of this style of implementation in a framework oriented around configuration (the first two paragraphs of https://docs.fluidproject.org/infusion/development/FrameworkConcepts.html#declarative-configuration may help to understand why Infusion considers this important) is that we can then write server configurations in pure JSON that simply specify which particular concrete grade to use. 

So deploying different sets of services can become a relatively simple matter - we just write new configurations. Main goals served: modularity, flexibility, customization, and making it easier for people to adapt the existing code in ways we didn't anticipate.

This approach is important to grasp (and can be a bit of a learning curve when starting with Infusion), so we can talk it through in more detail to make sure it's clear. It took me some time to grasp it, so don't worry if that's your experience as well.

Reply:
Yes, it is a bit difficult to get my head around this.
Although I understand some of the benefits, that you mentioned, of using such code structure, but I am not completely clear about how we will be going around with it.
I guess there will be a need to talk about this in more detail, as you mentioned.
One question to start with - What do you mean by defining `public API` in the first grade above?

Reply:
Sorry, I realize that's confusing in the context of this project - in this case I simply mean the common functions (exposed via Invoker definitions) that any integrating code could expect to work with.

I'll see later today if I can write a more concrete example to make the approach clearer, but generally wherever possible we want to implement polymorphic behaviour and similar object-oriented features through Infusion's built-in features like Invokers. It's less obviously useful when starting out, but it make a large difference as a codebase grows more complex.

Reply:
So, the utility functions (like `dictionary()` and `checkDictionaryErr()`), that are common to all the handlers of the dictionary service, would go in here?

Reply:
If they'd be shared functionality, they could be gathered under the abstract grade. It's also possible you'll simply want to define those common functions as Invokers on the grade and have concrete implementations in other grades.

Generally you should name function name to associate it with the relevant grade. So something like `adaptiveContentServices.dictionary.checkDictionaryErr` rather than a straight function name.

Reply:
Sounds good! I think another goal should be to break the functions down into smaller pieces and avoid switch-style "branching if" statements - using polymorphism with Invokers should help here.

Reply:
Okay! I will try this too, in the next commit.

Reply:
I have tried refactoring based on these lines.
However, I don't think I have correctly broken down the function to smaller pieces.